### PR TITLE
Implement SpatialPyramidPooling

### PR DESCRIPTION
**Pull request by yardstick17**
PR Inspired from the work of @luizgh and @RikHeijdens.
References:
- He, Kaiming et al (2015): Spatial Pyramid Pooling in Deep Convolutional Networks for Visual Recognition. http://arxiv.org/pdf/1406.4729.pdf.
- [Implement SpatialPyramidPooling](https://github.com/Lasagne/Lasagne/pull/799)

### CODE REVIEW

**Comment by sb2nov** [comment]
Jenkins, test this please.

File: tensorflow/python/layers/pooling.py [hunk 1]
```diff
@@ -416,6 +420,139 @@ def max_pooling2d(inputs,
...
+      result.append(pool_result)
+  return result
+
+
+def spatial_pyramid_pooling(inputs, dimensions=None,
+                            mode='max', implementation='kaiming'):
+  """
+    Spatial pyramid pooling (SPP) is a pooling strategy to result in an output of fixed size.
+    It will turn a 2D input of arbitrary size into an output of fixed dimension.
+    Hence, the convlutional part of a DNN can be connected to a dense part
```
Comment (RikHeijdens) [inline tensorflow/python/layers/pooling.py]: Type: convlutional -> convolutional

**Comment by fchollet** [comment]
Maybe this would be a better fit for contrib. The use case for this layer is fairly niche and wouldn't warrant inclusion in core layers.

A bit of feedback:

- It is preferable not to have API keywords related to implementation details. API keywords should be about domain-specific parameters, i.e. about what users care about.
- The API proposed is not consistent with that of other pooling layers.

**Comment by yardstick17** [comment]
@fchollet
- [x] Moved the layer to contrib.
- [x] Implementation based (domain-specific) parameters.
- [ ] The spp-layer API is not consistent with the other layer API because it's the uniqueness of its definition as per my understanding. Please point me if I can make it consistent with the existing layer APIs.

**Comment by yardstick17** [comment]
@fchollet I have updated the PR, please have a look.

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  """
+  layer = SpatialPyramidPooling(dimensions=dimensions,
+                                mode=mode,
+                                implementation=implementation)
+  return layer.apply(inputs)
+
+
+class SpatialPyramidPooling(base.Layer):
+  """
+    Spatial pyramid pooling (SPP) is a pooling strategy to result in an output of fixed size.
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Invalid docstring format (should start with a one-line summary, not a line break)

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    msg = "Mode must be either 'max' or 'avg'. Got '{0}'"
+    raise ValueError(msg.format(mode))
+
+  result = []
+  n = pool_dimension
+  for row in range(pool_dimension):
+    for col in range(pool_dimension):
+      # start_h = floor(row / n * h)
+      start_h = math_ops.cast(
+        math_ops.floor(math_ops.multiply(math_ops.divide(row, n), math_ops.cast(h, dtypes.float32))), dtypes.int32)
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Style: many lines in the PR are too long, including in docstrings.

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.layers import base
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import nn
+
+
+def max_pool_2d_nxn_regions(inputs, pool_dimension, mode):
+  """
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Style: docstrings should start with a 1-line summary.

**Review by fchollet — COMMENTED** [review]
Thank you for your patience.

I've made some quick comments on the PR but they are not exhaustive. I have not reviewed the PR in detail.

You state in the code that the code is ported from Lasagne, which is under a different license. For legal reasons, if you want to move ahead with this PR, you would need to "reimplement the functionality from scratch" rather than "porting it", which means no copy/paste and no reference to Lasagne.

**Comment by martinwicke** [comment]
@fchollet can you take another look?

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        spatial_bin_dimensions: The list of dimensions define the output dimension
+          of each pooling level. The value of each dimenstion is the level of
+          the spatial pyramid.
+        mode: Pooling mode 'max' or 'avg'.
+  """
+
+  def __init__(self, spatial_bin_dimensions=None, mode='max', **kwargs):
+    super(SpatialPyramidPooling, self).__init__(**kwargs)
+    self.mode = mode
+    self.spatial_bin_dimensions = spatial_bin_dimensions if spatial_bin_dimensions is not None else [4, 2, 1]
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Make line shorter (by having an explicit `if` block)

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    for bin_dimension in self.spatial_bin_dimensions:
+      pool_list += self.__spatial_pooling_in_bins(inputs, bin_dimension)
+    return array_ops.concat(values=pool_list, axis=1)
+
+  def _compute_output_shape(self, input_shape):
+    num_features = sum(p * p for p in self.spatial_bin_dimensions)
+    return tensor_shape.TensorShape([None, input_shape[0] * num_features])
+
+
+  def __spatial_pooling_in_bins(self, inputs, bin_dimension):
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Private methods start with a single `_`

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    pool_list = []
+    for bin_dimension in self.spatial_bin_dimensions:
+      pool_list += self.__spatial_pooling_in_bins(inputs, bin_dimension)
+    return array_ops.concat(values=pool_list, axis=1)
+
+  def _compute_output_shape(self, input_shape):
+    num_features = sum(p * p for p in self.spatial_bin_dimensions)
+    return tensor_shape.TensorShape([None, input_shape[0] * num_features])
+
+
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Remove blank line

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  def __spatial_pooling_in_bins(self, inputs, bin_dimension):
+    """
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Missing one-line docstring summary

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    Args:
+      inputs: The tensor over which to pool. Must have rank 4.
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: I don't see any check w.r.t. assumptions about `inputs`. Consider setting a `self.input_spec` attribute in `__init__`.

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  return layer.apply(inputs)
+
+
+class SpatialPyramidPooling(base.Layer):
+  """Spatial pyramid pooling (SPP) is a pooling strategy to result in an output of fixed size.
+
+    Arguments:
+        spatial_bin_dimensions: The list of dimensions define the output dimension
+          of each pooling level. The value of each dimenstion is the level of
+          the spatial pyramid.
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Sorry, this description is confusing, please clarify. Consider that most people reading this docstring come to it with no prior context.

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+  Returns:
+    Output tensor.
+  """
+  layer = SpatialPyramidPooling(spatial_bin_dimensions=spatial_bin_dimensions, mode=mode)
+  return layer.apply(inputs)
+
+
+class SpatialPyramidPooling(base.Layer):
+  """Spatial pyramid pooling (SPP) is a pooling strategy to result in an output of fixed size.
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: Please expand on the description, this is not sufficient to understand what the layer does. Put yourself in the shoes of a user: why/when should they use this layer?

File: tensorflow/contrib/layers/python/layers/pooling.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+    Arguments:
+        spatial_bin_dimensions: The list of dimensions define the output dimension
+          of each pooling level. The value of each dimenstion is the level of
+          the spatial pyramid.
+        mode: Pooling mode 'max' or 'avg'.
+  """
+
+  def __init__(self, spatial_bin_dimensions=None, mode='max', **kwargs):
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/pooling.py]: `mode` is not a very god API. `pooling_mode`  is better (explicit).

`bin_dimensions` would also be better than `spatial_bin_dimensions` (unless the bin dimensions could potentially not be spatial).

**Comment by fchollet** [comment]
Thank you for your patience. Sorry for the delayed review.

**Comment by drpngx** [comment]
@yardstick17 did you address all of @fchollet comments?

File: tensorflow/contrib/layers/python/layers/pooling.py (hunk 11 above)
Comment (yardstick17) [reply tensorflow/contrib/layers/python/layers/pooling.py]: Done.

**Comment by yardstick17** [comment]
@drpngx Yes, I incorporated the changes suggested by @fchollet.

**Comment by drpngx** [comment]
Thanks! @fchollet it's ready to review again.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
@fchollet can you see if the changes are good?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by Magisun** [comment]
@fchollet any updates on this?

**Review by fchollet — COMMENTED** [review]
Sorry for the delayed review.

I am not sure there is much a point in merging a new layer into `contrib.layers` at this point. It may be better for yourself and for other users to simply close this PR and ask you to maintain this layer on your own as a TF extension repository (which others could search for, find, and install, in order to use this layer in their models). I don't really see ourselves maintaining it in the future.

This would be my preferred course of action. Other than than, the code looks good, although it features style issues (such as overly long lines).

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: It has been 30 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: It has been 45 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: It has been 60 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: It has been 75 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 96 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@yardstick17 are you OK with the solution @fchollet is proposing?

**Comment by martinwicke** [comment]
I am sorry about the back and forth, and the long wait. We are very aware that the reviews are a bottleneck, and we are working on ways to remove that bottleneck. We also cannot maintain and support everything in TensorFlow, so for now, your own repo is probably the best option. 

We are working improving how the community can contribute to TensorFlow. One such approach is a new SIG, maintaining a new repo, containing layers that are important, but which we cannot maintain in TensorFlow itself. A sketch of this proposal is in https://github.com/tensorflow/community/pull/18, expect a more detailed proposal soon as we form a special interest group.

I will close this PR.

**Pull request closed without merging**